add ability to specify multiple grok patterns

talevy opened the pull request:
now you can specify a list of grok patterns to match your field with
and the first one to successfully match wins.

Fixes #17903.

spinscale commented on docs/reference/ingest/ingest-node.asciidoc:
```diff
@@ -1033,7 +1033,7 @@ Grok expression.
 | Name                   | Required  | Default             | Description
 | `field`                | yes       | -                   | The field to use for grok expression parsing
-| `pattern`              | yes       | -                   | The grok expression to match and extract named captures with
+| `patterns`             | yes       | -                   | An ordered list of grok expression to match and extract named captures with. Returns on first pattern in the list that matches.
```
expressions

talevy replied:
thanks!

spinscale commented on modules/ingest-grok/src/main/java/org/elasticsearch/ingest/grok/GrokProcessor.java:
```diff
@@ -77,20 +84,26 @@ public Factory(Map<String, String> builtinPatterns) {
-                grok = new Grok(patternBank, matchPattern);
-            } catch (Exception e) {
-                throw newConfigurationException(TYPE, processorTag, "pattern", "Invalid regex pattern. " + e.getMessage());
+            List<Grok> groks = new ArrayList<>();
```
can be initialized with the same size as `matchPatterns`?

talevy replied:
no reason it can't! I'll update

spinscale commented on modules/ingest-grok/src/test/java/org/elasticsearch/ingest/grok/GrokProcessorFactoryTests.java:
```diff
@@ -59,42 +59,54 @@ public void testBuildMissingField() throws Exception {
+            fail("should fail");
+        } catch (ElasticsearchParseException e) {
+            assertThat(e.getMessage(), equalTo("[patterns] List of patterns must not be empty"));
+        }
```
you could rewrite this as

```
ElasticsearchParseException e =  expectThrows(ElasticsearchParseException.class, () -> factory.create(config));
assertThat(e.getMessage, ... );
```

talevy replied:
oh that's a neat little helper. Is this meant to replace all uses of the old pattern? If so, I would think it would make sense to update all of the test cases, not just this one. what do you think?

spinscale replied:
++, it's a java 8 only idiom, which is not a problem for ingest, no backports

talevy replied:
updated

spinscale commented:
Is there a possibility to find out which match actually matched for debugging/simulating?

Aside from that LGTM

talevy commented:
> Is there a possibility to find out which match actually matched for debugging/simulating?

Like a special debug section that would be added to the response of `simulate?verbose` responses?

spinscale commented:
I havent spent too much thoughts into it, how to do it, I just figured that this might be important for debugging. Simulate API w/ verbose sounds about right to me. That does not need to be part of this PR however!

rjernst commented:
This feature (multiple groks) seems trappy. It operates linearly on the number of patterns. Instead, I think it would be better to fix grok to be able to have multiple alternative patterns (so it can be compiled to a single regex)?

talevy commented:
@rjernst, true. It would be more performant to scan through the text once. I guess I was looking for a "quick" fix here and was not too worried about performance for these less frequent multi-regex cases.

talevy commented:
@rjernst, after thinking about it for a little bit. what would your expected behavior be for when multiple expressions match? how would you deterministically decide on a winner without this trappiness of order?

clintongormley commented:
> Instead, I think it would be better to fix grok to be able to have multiple patterns in a single grok (so it can be compiled to a single regex)?

This could work as long as users only use named captures.  If they use positional captures then stringing regexes together will break.

talevy commented:
@clintongormley you're right. We are only interested in non-overlapping named-capture matches. so a simple union will do the trick.

@rjernst I have updated this to combine the patterns into one grok expression using `|`

@spinscale As part of this, I have introduced a `trace_match` flag for debugging which pattern matched. When set, the GrokProcessor will inject the index into the `patterns` list of the pattern that matched. The field is `_ingest._grok_match_index`. The combined regex expression gets rather complicated with this flag turned on because I introduce a fake `(?<_ingest._grok_match_index.IDX>^).*` pattern in front of each pattern to match the beginning of that pattern.

for example, this would be injected for a patterns list whose second pattern was matched.

```
_ingest._grok_match_index.0 => ""
```

Then I update this value to be

```
_ingest._grok_match_index => "0"
```

upon exiting from the processor, so this is what people actually see.

does that make sense?


A side-effect of multiple patterns that can match is that all the non-matched pattern's fields were inserted 
into the document as `null` values. This polluted the document with unrelated fields. In the latest revision, I filter these out. So, if a user expected a null match in their regex, I will simply not insert that field.

^ this kind of bothers me, but I am not sure how to remedy this at the moment

talevy commented:
```
org.elasticsearch.smoketest.IngestWithMustacheIT.test {yaml=ingest/20_combine_processors/Test logging}
```
is failing. I believe this is due to he `null` value filtering. will investigate further

clintongormley commented:
> The combined regex expression gets rather complicated with this flag turned on because I introduce a fake (?<_ingest._grok_match_index.IDX>^).* pattern in front of each pattern to match the beginning of that pattern.

Nice solution!

> In the latest revision, I filter these out. So, if a user expected a null match in their regex, I will simply not insert that field.  this kind of bothers me, but I am not sure how to remedy this at the moment

Isn't it just a matter of only filtering out null matches for names beginning with `_grok_match_index`?

talevy commented:
@clintongormley 
when `trace_match == false` then I do introduce meta name-captures, I only wrap them in an 
unnamed group. So, I cannot differentiate between the two. I thought it would be more performant that way, since I do not care about tracking. I will do some benchmarking on this to find out if this is _actually_ more performant.

talevy commented:
I don't think there is a way to distinguish between the different fields without prefixing each recursive named capture with the tracking `_ingest._grok_match_index.IDX`. should I do this?

talevy commented:
ping for followup

martijnvg commented on modules/ingest-grok/src/main/java/org/elasticsearch/ingest/grok/GrokProcessor.java:
```diff
@@ -58,14 +77,44 @@ public String getType() {
     }
 
+    // TODO(talevy): cleanup with builders? make this a Collector?
+    private static String combinePatterns(List<String> patterns, boolean traceMatch) {
```
Can we add a unit test for this method that tests the combining logic in isolation?

talevy replied:
adding now

martijnvg commented:
Left one test related comment. Aside from that LGTM. I like the pattern combining logic!

talevy commented:
@martijnvg, thanks!

what are your thoughts on this:

> In the latest revision, I filter these out. So, if a user expected a null match in their regex, I will simply not insert that field. this kind of bothers me, but I am not sure how to remedy this at the moment

beforehand we would store all named captures as fields... even if they were null. Now that _all_ named captures (from different regexes) are in one expression, I cannot really distinguish between them, so I went with not inserting any of them if they were null.

martijnvg commented:
> beforehand we would store all named captures as fields... even if they were null. Now that all named captures (from different regexes) are in one expression, I cannot really distinguish between them, so I went with not inserting any of them if they were null.

Discussed this with @talevy and we think it is okay to ignore null keys in the ingest document if the grok parser can't match the named capture. (before we would add keys with null values)

talevy commented:
@martijnvg added unit tests for `combinePatterns`

martijnvg commented on modules/ingest-grok/src/main/java/org/elasticsearch/ingest/grok/GrokProcessor.java:
```diff
@@ -58,14 +77,43 @@ public String getType() {
     }
 
-    String getMatchField() {
+    public String getMatchField() {
```
I think these getters can remain to be package protected?

talevy replied:
yup. on it

martijnvg commented:
Left one minor comment. LGTM
